Building your own MCP tools (#356)
neo4j-graphacademy/courses: asciidoc/courses/genai-mcp-neo4j-tools/modules/1-what-is-mcp/lessons/1-getting-started/lesson.adoc

= What is MCP?

The Model Context Protocol, commonly referred to as MCP, is an open protocol that enables agents and AI applications to work together in a standardized way.

> Before we dive into MCP, let's familiarise ourselves with the concept of an **Agent**.

Reviewer: ```suggestion
Before learning more about MCP, its important to understand the concept of an **Agent**.
```

== What are Agents?

> There are many defintions of an Agent, but for the context of this course, an Agent is a system that **acts independently** to **achieve a specific goal**.

Reviewer 2: using tools to access information and take action

> A common type of framework for building agents is the **ReAct** framework.

Reviewer: ```suggestion
A common type of framework for building agents is the **ReAct**  (reason act) framework.
```

When a ReAct agent receives an input, it will use a Large Language Model (LLM) to **reason** on the problem at hand, and use the descriptions of tools to select the correct tool(s) to solve the problem, and **act** by calling the tools, either in sequence or in parallel, to perform actions.

The agent will continue to reason and act until the goal is achieved.

An example of a tool may be a function that searches the web, calls an external API or uploads a file to a cloud storage service.

== Where does MCP fit in?

> While agents may have access to tools within the scope of their application, MCP provides a standardized way for the agent to consume tools from other providers and either hosted remotely or accessed locally through Standard Input/Output (stdio).

Reviewer 2: I would leave the transport protocol for later, 
> within the scope of their application -> perhaps within the code? not sure if people understand what scope means here

== Architecture

The MCP protocol consists of the following elements.

=== Servers

> The protocol follows a client-server archicture, similar to microservices, where **Servers** provide capabilities through a list of **tools**.

Reviewer 2: tools, resources, prompts (today)

> Each tool that the server hosts has additional metadata assigned to it, including a unique identifier, a description of when that tool can be used, and a list of parameters that the tool can receive.

Reviewer 2: explanation of parameters and outputs (new spec)
